python/cryptol/bv.py

```diff
@@ -0,0 +1,385 @@
+
+    def size(self) -> int:
+        """Width of the BV (i.e., the "bit width" of the value)."""
+        return self.__size
+
+    def value(self) -> int:
+        """The unsigned integer interpretation of the BV."""
+        return self.__value
+
+    def __concat_single(self, other : 'BV') -> 'BV':
+        if isinstance(other, BV):
+            return BV(self.__size + other.__size, (self.__value << other.__size) + other.__value)
+        else:
+            raise ValueError(f'Cannot concat BV with {other!r}')
+
+    def concat(self, *args : List['BV']) -> 'BV':
+        """Concatenate the given BV to the right of `self`."""
+        return reduce(lambda acc, b: acc.__concat_single(b), args, self)
+    
+    @staticmethod
+    def join(*bs : List['BV']) -> 'BV':
+        """Concatenate the given BV in order."""
+        return reduce(lambda acc, b: acc.__concat_single(b), bs, BV(0,0))
+
+    def zero(self) -> 'BV':
+        """Return the zero value for this BV' size."""
+        return BV(self.size() ,0)
+
+    def to_int(self) -> 'BV':
+        """Return the unsigned integer the BV represent."""
+        return self.__value
+
+    def to_signed_int(self) -> int:
+        """Return the signed (i.e., two's complement) integer the BV represents."""
+        if self.__value >= 0:
+            return self.__value
+        else:
+            # negOfNat (2^n - x.to_nat)
+            return 0 - ((2 ** self.__size) - self.__value)
+
+    def __eq__(self, other : Any) -> bool:
+        if isinstance(other, BV):
+            return self.__size == other.__size and self.__value == other.__value
+        else:
+            return False
+
+    def __index__(self):
+        return self.__value
+
+    def __len__(self):
+        return self.__size
+
+    def __bytes__(self):
+        byte_len = (self.__size // 8) + (0 if self.__size % 8 == 0 else 1)
+        return self.__value.to_bytes(byte_len, 'big')
+
+
+    def split(self, size : int) -> List['BV']:
+        """Split the BC into a list of BVs of length `size`.
+        
+        Note: `self.size()` must be divisible by `size`."""
+        if not isinstance(size, int) or size <= 0:
+            raise ValueError(f'`size` argument to splits must be a positive integer, got {size!r}')
+        if not self.size() % size == 0:
+            raise ValueError(f'{self!r} is not divisible into equal parts of size {size!r}')
+        mask = (1 << size) - 1
+        return [BV(size, (self.__value >> (i * size)) & mask)
+                for i in range(self.size() // size - 1, -1, -1)]
+
+
+    def popcount(self) -> int:
+        """Return the number of bits set to `1` in `self`."""
+        return bin(self).count("1")
+
+    @staticmethod
+    def from_bytes(bs : bytes, *, size=None, byteorder='big') -> 'BV':
+        """Convert the given bytes into a BV.
+
+        Resulting BV has length `len(bs) * 8.` 
+        
+        `size` keyword argument specifies BV's size (and
```

Comment:
It would be preferable to use the Sphinx format for the docstring, especially when it comes to enumerating arguments: https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html

This is because we extract the docstrings and show them in the generated HTML docs.

Reply:
See https://github.com/GaloisInc/argo/pull/116/commits/3cf029b7b1111e484370b7d7b2cf46f15fbed5bf